x/auth/ante/sigverify.go

```diff
@@ -0,0 +1,342 @@
+			panic(err)
+		}
+		isd.ak.SetAccount(ctx, acc)
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// ValidateSigCountDecorator takes in Params and returns errors if there are too many signatures in the tx for the given params
+// otherwise it calls next AnteHandler
+// Use this decorator to set parameterized limit on number of signatures in tx
+// CONTRACT: Tx must implement SigVerifiableTx interface
+type ValidateSigCountDecorator struct {
+	ak keeper.AccountKeeper
+}
+
+func NewValidateSigCountDecorator(ak keeper.AccountKeeper) ValidateSigCountDecorator {
+	return ValidateSigCountDecorator{
+		ak: ak,
+	}
+}
+
+func (vscd ValidateSigCountDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	sigTx, ok := tx.(SigVerifiableTx)
+	if !ok {
+		return ctx, sdkerrors.Wrap(sdkerrors.ErrTxDecode, "Tx must be a sigTx")
+	}
+
+	params := vscd.ak.GetParams(ctx)
+	pubKeys := sigTx.GetPubKeys()
+
+	sigCount := 0
+	for _, pk := range pubKeys {
+		sigCount += types.CountSubKeys(pk)
+		if uint64(sigCount) > params.TxSigLimit {
+			return ctx, sdkerrors.Wrapf(sdkerrors.ErrTooManySignatures,
+				"signatures: %d, limit: %d", sigCount, params.TxSigLimit)
+		}
+	}
+
+	return next(ctx, tx, simulate)
+}
+
+// DefaultSigVerificationGasConsumer is the default implementation of SignatureVerificationGasConsumer. It consumes gas
+// for signature verification based upon the public key type. The cost is fetched from the given params and is matched
+// by the concrete type.
+func DefaultSigVerificationGasConsumer(
+	meter sdk.GasMeter, sig []byte, pubkey crypto.PubKey, params types.Params,
+) error {
+	switch pubkey := pubkey.(type) {
+	case ed25519.PubKeyEd25519:
+		meter.ConsumeGas(params.SigVerifyCostED25519, "ante verify: ed25519")
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidPubKey, "ED25519 public keys are unsupported")
+
+	case secp256k1.PubKeySecp256k1:
+		meter.ConsumeGas(params.SigVerifyCostSecp256k1, "ante verify: secp256k1")
+		return nil
+
+	case multisig.PubKeyMultisigThreshold:
+		var multisignature multisig.Multisignature
+		codec.Cdc.MustUnmarshalBinaryBare(sig, &multisignature)
+
+		consumeMultisignatureVerificationGas(meter, multisignature, pubkey, params)
+		return nil
+
+	default:
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidPubKey, "unrecognized public key type: %T", pubkey)
+	}
+}
+
+func consumeMultisignatureVerificationGas(meter sdk.GasMeter,
+	sig multisig.Multisignature, pubkey multisig.PubKeyMultisigThreshold,
+	params types.Params) {
+	size := sig.BitArray.Size()
+	sigIndex := 0
+	for i := 0; i < size; i++ {
+		if sig.BitArray.GetIndex(i) {
+			DefaultSigVerificationGasConsumer(meter, sig.Sigs[sigIndex], pubkey.PubKeys[i], params)
+			sigIndex++
+		}
+	}
+}
+
+// Internal function that simulates gas consumption of signature verification when simulate=true
+// TODO: allow users to simulate signatures other than auth.StdSignature
```

Comment:
Not quite sure how we should do this

Reply:
merge in master, it should be disabled now

Reply:
I don't understand, can you clarify?